[SYCL][DOC] Add ext_intel_virtual_functions specification (#10540)
intel/llvm: sycl/doc/extensions/proposed/sycl_ext_intel_virtual_functions.asciidoc

= sycl_ext_oneapi_virtual_functions

:source-highlighter: coderay
:coderay-linenums-mode: table

// This section needs to be after the document title.
:doctype: book
:toc2:
:toc: left
:encoding: utf-8
:lang: en
:dpcpp: pass:[DPC++]

// Set the default source code type in this document to C++,
// for syntax highlighting purposes.  This is needed because
// docbook uses c++ and html5 uses cpp.
:language: {basebackend@docbook:c++:cpp}

== Notice

[%hardbreaks]
Copyright (C) 2023-2023 Intel Corporation.  All rights reserved.

Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
permission by Khronos.

== Contact

To report problems with this extension, please open a new issue at:

https://github.com/intel/llvm/issues

== Dependencies

> This extension is written against the SYCL 2020 revision 7 specification.  All

Reviewer: This document should presumably be rebased on revision 8 now.

SYCL specification refer to that revision.

This extension also depends on the following other SYCL extensions:

* link:../experimental/sycl_ext_oneapi_kernel_properties.asciidoc[
  sycl_ext_oneapi_kernel_properties]
* link:../experimental/sycl_ext_oneapi_properties.asciidoc[
  sycl_ext_oneapi_properties]
* link:../experimental/sycl_ext_oneapi_named_sub_group_sizes.asciidoc[
  sycl_ext_oneapi_named_sub_group_sizes]

== Status

This is a proposed extension specification, intended to gather community
feedback.  Interfaces defined in this specification may not be implemented yet
or may be in a preliminary state.  The specification itself may also change in
incompatible ways before it is finalized.  *Shipping software products should
not rely on APIs defined in this specification.*

== Backend support status

The APIs in this extension may be used only on a device that has
`aspect::ext_oneapi_virtual_functions`.  The application must check that the
device has this aspect before submitting a kernel using any of the APIs in this
extension.  If the application fails to do this, the implementation throws
a synchronous exception with the `errc::kernel_not_supported` error code

> when the kernel is submitted to the queue.

Author: My assumption was that FPGA devices are unlikely to support this feature

be used:

- if an object is constructed in host code, calling a virtual member function
  for that object in device code has undefined behavior;
- if an object is constructed in device code on a device `A`, calling a virtual
  member function for that object in host code, or on another device `B` has
  undefined behavior;

=== New properties

Due to the indirect nature of virtual member functions, compiler may not be able
to understand which exact virtual member function is being called (i.e. which
class it belongs to).  Moreover, when code is distributed over several
translation units, compiler may not be able to even see definitions of all
virtual member functions which may be called in a kernel.

Therefore, to provide a mechanism for an implementation to help detect virtual
member functions, which are going to be used from kernels and enforce necessary
restrictions on them, new compile-time-constant properties are proposed.

[source,dpcpp]
----
namespace sycl::ext::oneapi::experimental {

  struct indirectly_callable_key {
    template <typename SetId>
    using value_t = property_value<indirectly_callable_key, SetId>;

>   };

Author: We've had a discussion with CFE folks and agreed that it is ok for CFE to parse properties for some special handling. This means that CFE can react to certain string literals passed down to `__sycl_detail__::add_ir_attribute_function` attribute for special handling (like converting them into attributes, emitting a diagnostic, etc.).

Therefore, the extension and design doc will be updated to follow option 2

>   };

Author: Both documents were updated to follow option to in 531083a82ccb48a2142bfcac6f8a71bd24813047

  struct calls_indirectly_key {

>     template <typename First = void, typename... SetIds>

Author: Applied in 683bc66e92e3acb4c652791cfd7a815e5afa5783

    using value_t = property_value<calls_indirectly_key, First, SetIds...>;
  };

  template <typename SetId = void>
  inline constexpr indirectly_callable_key::value_t<SetId> indirectly_callable;

  template <typename First = void, typename... Rest>
  inline constexpr calls_indirectly_key::value_t<First, Rest...>
      calls_indirectly;

  template <>
  struct is_property_key<indirectly_callable_key> : std::true_type {};
  template <> struct is_property_key<calls_indirectly_key> : std::true_type {};
}
----

> Before describing those properties those properties in more detail, a couple of

Reviewer 2: ```suggestion
Before describing those properties in more detail, a couple of
```

Author: Applied in 683bc66e92e3acb4c652791cfd7a815e5afa5783

new terms are introduced to simplify the extension specification:

Set of virtual member functions:: a group of virtual member functions which are

> defined with `indirectly_callable` property with the same value of the property

Reviewer: ```suggestion
defined with the `indirectly_callable` property and with the same value of the property
```

Author: Applied in 683bc66e92e3acb4c652791cfd7a815e5afa5783

> parameter `SetId`. For the simplicity, this will also be further refferred to as

Reviewer 2: ```suggestion
parameter `SetId`. For simplicity, this will also be further referred to as
```

a _set_, or as a _set of virtual functions_.

Kernel declares a use of a set of virtual member functions:: a kernel is
considered to be declaring a use of a set of virtual member functions `SetIdA`
when it is submitted with `calls_indirectly` property with `SetIdA` included
into the property parameter `SetIds`. If `SetIdA` is not included into the
property parameter `SetIds`, or if a kernel is submitted without the property,

> then it is *not* considered to be declaring a use of the set of virtual member

Reviewer 3: > Good question. That actually may impact caches that we have for programs, because at least for in-memory cache we don't include properties as a key.

I think this is a general issue with kernel properties that is not specific to `calls_indirectly`.  According to [sycl_ext_oneapi_kernel_properties](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc), the same kernel can be submitted multiple times with different properties.  Therefore, I assume that the caching mechanism would have to include the set of kernel properties as part of the cache key.

However, I notice that @Pennycook did *not* allow the same kernel to be submitted multiple times with different properties with the new APIs in [sycl_ext_oneapi_enqueue_functions](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc).  Instead, a kernel can only have properties if the kernel is defined as a named kernel object.  Thus, the properties become part of the kernel's type.  If this is the new direction, maybe we should consider changing [sycl_ext_oneapi_kernel_properties](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc), so that it has the same restriction as [sycl_ext_oneapi_enqueue_functions](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc).  This would mean that the kernel properties would *not* need to be part of the cache key.

> The idea was that the call can't be done at runtime, but it is ok to have it statically in a kernel.

I think we want to allow a compiler to emit a diagnostic if it can prove that a call site will always go to a virtual function that is not in the kernel's set of allowed virtual functions.  In most cases, I think the compiler will not be able to prove this, but in some cases it might be possible.

> then it is *not* considered to be declaring a use of the set of virtual member

Reviewer 2: > If this is the new direction, maybe we should consider changing [sycl_ext_oneapi_kernel_properties](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc), so that it has the same restriction as [sycl_ext_oneapi_enqueue_functions](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc).

I think this makes sense.

This new direction was in part informed by some of the launch queries and forward progress stuff that we drafted after the initial version of sycl_ext_oneapi_kernel_properties.  If a kernel can only ever have one set of properties, querying launch limits is much simpler; if the same kernel can be launched with multiple compile-time properties, those properties have to be provided at query time (always painful, sometimes impossible).

I hadn't updated the old specification because there were so many moving parts, and I wanted to wait for the new direction to settle.  If there are no objections to this simplification, I think we should do it.

functions.

|===
|Property|Description
|`indirectly_callable`
|The `indirectly_callable` property indicates that a virtual member function is
a device function, thus making it available to be called from SYCL kernel and
device functions. Should only be applied to virtual member functions and to do

> so, function-style `SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY` macro should

Reviewer 2: Why are we using this instead of the mechanism defined in the kernel properties extension?  That would give:

```
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((sycl::indirectly_callable<SetId>))
```

...and would allow multiple function properties to be defined in one place.

Reviewer 2: Won't the FE have to implement that sort of parsing to handle compile-time properties eventually, anyway?  I'm trying to work out if this is just going to impact this extension (and #11861) or if you think there's a fundamental issue with the `SYCL_EXT_ONEAPI_FUNCTION_PROPERTY` design.

Author: > Won't the FE have to implement that sort of parsing to handle compile-time properties eventually, anyway?

Well, we are trying to avoid this for as long as possible :)

If I understand correctly, compile-time properties were introduced to move away from attributes with an idea that less compiler changes would be needed to implement a new property. We are still not in a state where introducing a new property can be solely done in headers, but we are not too far away from that for simple properties which are expected to trigger generation of certain SPIR-V decoration and nothing more.

Front-end is not that modular as the rest of the toolchain and any modifications to it are more likely to be scattered around the whole codebase instead of being isolated into a single SYCL-specific component like it is done for middle-end part of the toolchain (new tools and compiler passes). This I think is one of reasons why we don't want to touch the front-end too often.

> if you think there's a fundamental issue with the `SYCL_EXT_ONEAPI_FUNCTION_PROPERTY` design.

I don't quite like properties on free functions: on the one hand we are trying to move away from attributes, but on the other hand we are anyway using them, but in a very restricting manner which results in those custom properties macro we are introducing for virtual functions and free function kernel extensions.

Properties as an intended library-only replacement for attributes look good and solid for kernels. For free functions, however, they do not look as much as a library-only thing and hence all the problems and conflicts we see: need to resort to preprocessor to attach properties and inability (or rather reluctance) to do FE semantic checking based on properties.

I wonder if it is possible to implement something based on trait specializations instead?

```c++
// SYCL headers:
template<auto &Func, typename SetId>
struct is_device_indirectly_callable {
private:
  using __annotated_func = __some_magic_builtin_to_attach_metadata(Func, "indirecly-callable", __builtin_sycl_unique_stable_name(SetId));
  int __check = { []() __attribute__((sycl_device)) {
      // we are turning a function into a device one
      // by calling it from another device function
      Func();
      // yeah, that's a problem to write a generic code which would make a call to an arbitrary (member) function
     } }, 42; // operator comma to disregard the lambda later
};

// User code:

struct Foo {
  virtual void bar() { /* ... */ }
};

struct set_A;

template <>
is_device_indireclty_callable<Foo::bar, set_A> {};
// specialization is not enough if there won't be an instantiation, unfortunately
```

Reviewer 2: Hm.  I don't think we considered traits for the function properties, and I'm not sure why.  I'll make some time to stare at this a bit more and see if there's any way that we could use traits as a general replacement here.  It would definitely be more library-only-friendly, and the macro has a lot of hacks I don't like (e.g., the double parentheses workaround for templated properties).

Reviewer 2: I think switching to traits would allow us to adopt similar syntax to the syntax for [embedding properties into a kernel](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc#embedding-properties-into-a-kernel):

```c++
namespace syclex = sycl::ext::oneapi::experimental;

// Default trait in the headers
namespace sycl::ext::oneapi::experimental {

template <auto& Function>
struct function_traits {
    auto get(properties_tag) {
        return properties{};
    }
};

}

...

// Declare some device function
void foo();

// Attach properties to the device function
template <>
struct syclex::function_traits<foo> {
    auto get(syclex::properties_tag) {
        return syclex::properties{syclex::sub_group_size<8>, syclex::indirectly_callable<>};
    }
};
```

If we go this route, I think it makes more sense to attach all the properties to the function at once rather than to have separate traits (like `is_indirectly_callable` for each property).  This will lead to fewer surprises in the long run.

There are a few ways we could make the syntax above a little simpler, but there'd be some downsides (like functions and kernels requiring different syntax).

```c++
// Pass properties directly (shown as a property list but could be a parameter pack)
// NB: Renamed to _properties because we can't easily extend this version with other traits
template <>
struct syclex::function_properties<foo,
    decltype(syclex::properties{syclex::sub_group_size<8>, syclex::indirectly_callable<>})>
{};
```

```c++
// Use some sort of helper to define the right properties interface inside function_traits
// We could still extend this with other traits
template <>
struct syclex::function_traits<foo>:
  syclex::helper<decltype(syclex::properties{syclex::sub_group_size<8>, sycl::indirectly_callable<>})>
{};
```

Tagging @rlandschulz and @gmlueck for comment too.

Reviewer 3: I think the traits solution won't work for `indirectly_callable` because our primary use case is to decorate a virtual member function.  The trait takes a non-type template parameter that is the address of the function, but you can only do this for functions declared at namespace scope and functions declared `static` in class scope.  Virtual functions are never static, so you would not be able to specialize the trait for these functions.

See here for an example: https://godbolt.org/z/3cj1reGfr

Reviewer 2: I really don't understand the non-type template parameter rules or virtual function rules enough to understand what we need to achieve here, but it may be worth noting that things compile if the trait takes a pointer to the member instead: https://godbolt.org/z/ffz5q3Wvf

Reviewer 3: I didn't realize you could do that!

I'm still not wild about the syntax, though.  The trait would have to be specialized after the function declaration, so it moves the property far away from the function to which it applies.

Reviewer: A pointer to a member function requires an object to obtain an address of an actual function since, in general, the dynamic type of the object is needed to resolve an address. It isn't clear to me from the discussion so far how the trait would be used. Since it isn't possible from library code to identify the calls that are actually made from within a kernel or other device function, no opportunity for diagnosing calls to functions with unmatched set IDs is possible without compiler support. If compiler support is required, then use of a trait would just increase complexity.

Author: Even though visually traits look better than macro, they do not look good at all with virtual member functions.

Also, as @tahonermann noted we will still need some compiler magic to do diagnostics and apply special handling for those functions - and I'm not sure if that would even be possible if those traits are never used by the runtime.

We have agreed that it is ok for the compiler to parse string literals passed to the properties internal attribute and therefore the doc was updated to use the generic properties interface in 531083a82ccb48a2142bfcac6f8a71bd24813047

I will mark this thread as resolved

be used. The macro accepts a single optional argument, which is passed to the
property parameter `SetId` and therefore has all the same requirements.

Optional parameter `SetId` specifies a set of virtual member functions this
function belongs to and at the same time it defines a group of kernels, which
can call this function, it must be a C++ typename. When the parameter is

> ommitted, a virtual member function is considered to belong to the default set.

Author: Applied in 683bc66e92e3acb4c652791cfd7a815e5afa5783

Calling a virtual member function from a kernel which does not declare use of a
set the virtual member function belongs to is an undefined behavior.

> |`calls_indirectly`

Reviewer 2: Bikeshedding: `calls_indirectly<SetA>` reads well, but I'm worried that `calls_indirectly` alone is confusing.

I'd prefer something like `indirect_calls` (i.e., "this kernel contains indirect calls") or `assume_indirect_calls` (i.e., "the compiler should assume this kernel contains indirect calls").

Author: Any other feedback here, @gmlueck maybe?

I'm fine with changing names. We can even have two properties: one that accepts `set` argument and another which doesn't, to avoid the need to add `<>` at the end.

It could be:
```
q.single_task(properties{assume_indirect_calls}, [](){}); // default set
q.single_task(properties{assume_indirect_calls_from<set_A, set_B>}, [](){}); // explicitly-provided sets
```

Author: > I presume we would also split `indirectly_callable`?

Right. Those 4 suggested properties looks good to me. @Pennycook, does it resolve your concern? If so, I will update the proposal

Reviewer 2: Yes, I think this is much clearer.

I think the last property should be called something like `indirectly_callable_in`, though.  The symmetry of to/from might imply that the "from" list should be a list of kernels, rather than a list of sets.

Author: > I think the last property should be called something like `indirectly_callable_in`, though.

`indirectly_callable_as`? "In" looks a bit weird to me

Reviewer 2: I don't think "as" is quite right, either.  It might help to look at some complete examples.

With "in", I think it's clear from the declarations below  that `foo` is in `set_one` and `set_two`, but `bar` is only in `set_one`.

```c++
virtual SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::indirectly_callable_in<set_one, set_two>)
void foo() {}

virtual SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::indirectly_callable_in<set_one>)
void bar() {}
```

With "as", the code below looks wrong to me.   Somebody unfamiliar with the extension might read this as setting up aliases: so `foo` is callable via the names `set_one` and `set_two`, whereas `bar` is callable via the name `set_one`.  Such a reader might then be confused about how these two functions can both be called "as" the same thing.
 
```c++
virtual SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::indirectly_callable_as<set_one, set_two>)
void foo() {}

virtual SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::indirectly_callable_as<set_one>)
void bar() {}
```

I'm not trying to argue that "in" is perfect, and I'd be very happy if we can come up with something better.  But I think "in" is better than "from" and "as".

Reviewer 3: Personally I still prefer "from".  I could live with "in", but I think "as" is confusing.

The words "to" and "from" are a natural pair.  They way I read the properties is:

* `assume_indirect_calls_to<set>` means "assume this kernel might call **to** any function in `<set>`"
* `indirectly_callable_from<set>` means "this function might be called **from** any kernel in `<set>`"

With "in", I guess would read the property as:

* `indirectly_callable_in<set>` means "this function is **in** the set of functions that might be called from kernels in `<set>`.

Reviewer 2: > The words "to" and "from" are a natural pair. They way I read the properties is:
> 
> `assume_indirect_calls_to<set>` means "assume this kernel might call to any function in `<set>`"
> `indirectly_callable_from<set>` means "this function might be called from any kernel in `<set>`"

This is exactly why I don't like "from".  Each  `<set>` contains functions, not kernels -- so what does it mean above for a function to be "called from any kernel in `<set>`"?

With "in" , I think the reading should be:

- `indirectly_callable_in<set>` means "This function is **in** this `<set>` of functions, and may be indirectly called from any kernel that targets this `<set>`".

This matches my understanding of what the extension is doing, at least, but I might have misunderstood something.

Reviewer 3: The way I see it, the "set" applies to both kernels and functions.  We can say that kernels are "in" a set, and we can say that functions are "in" a set.  However, these are both notional concepts.  The implementation doesn't require all functions in the same set to be bundled together into the same device image, for example.  It's just a convenient way to indicate which kernels could make calls to which functions.  The implementation assumes that any kernel in set _A_ might indirectly call any function that is in set _A_.

Reviewer 2: That might work as a solution, but it's not what the specification currently says.  It says:

> For that the pass should compile a list of aspects used by each set of
indirectly callable functions (as defined by `indirectly_callable` property set
by user) and then append those aspects to every kernel which use those sets (as
defined by `calls_indirectly` property set by user).

...which I read as saying that each set is a set of functions, not functions and kernels.

I think if we want to use to/from then the rest of the wording would need to be updated to say explicitly that each set contains a mix of functions and kernels, and that any kernel in a set might indirectly call any function in that set.

Reviewer 3: That quote comes from the design document, not the spec, right?

Reviewer 2: Oops.  Yes, you're right -- sorry, I was searching for "set" and didn't realize there were multiple documents on the page.

But the specification contains similar wording about the set containing only functions:

> Set of virtual member functions:: a group of virtual member functions which are
defined with the `indirectly_callable` property and with the same value of the
property parameter `SetId`. For simplicity, this will also be further referred
to as a _set_, or as a _set of virtual functions_.
> 
> Kernel declares a use of a set of virtual member functions:: a kernel is
considered to be declaring a use of a set of virtual member functions `SetIdA`
when it is submitted with `calls_indirectly` property with `SetIdA` included
into the property parameter `SetIds`. If `SetIdA` is not included into the
property parameter `SetIds`, or if a kernel is submitted without the property,
then it is *not* considered to be declaring a use of the set of virtual member
functions.

Reviewer 3: I think we could adjust the wording in the specification to align with the names we choose.  What is the easiest mental model to understand?  Is it the "to/from" model or the "to/in" model?

Reviewer 2: Sorry for the delayed response.  I'm still leaning towards Alexey's original model of "here are the functions that a kernel can call" -- I think it will be simpler for developers to understand, and it's more aligned with the restrictions.

I still find it weird that a set could contain both functions and kernels.  It just feels a little messy, as abstractions go...

If `kernel_one` contains indirect calls to `foo` and `bar`, it's easy to explain that `kernel_one` includes calls to the sets containing `foo` and `bar`.  Saying `kernel_one` is also a member of those sets feels like a stretch to me.  Even more so if `kernel_two` also contains indirect calls to `foo` and `baz`: that would give us the sets {`kernel_one`, `kernel_two`, `foo`}, {`kernel_one`, `bar`}, {`kernel_two`, `baz`}?  But is there any meaning to the fact that `kernel_one` and `kernel_two` are sometimes in the same set?

If the issue here is really with to/in, maybe we could make the concept of a function set a little more general?

```c++
// Declare that foo is a member of function sets one and two
virtual SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::function_set<set_one, set_two>)
void foo() {}

// Declare that this kernel can make indirect calls to sets one and two
q.single_task(properties{assume_indirect_calls_to<set_one, set_two>}, [](){});
```

This avoids having to decide between "from" and "in", but keeps "to" (which we seem to agree has the correct semantics).  As a bonus, we can potentially re-use the concept of a group of device functions for other extensions.

Reviewer 3: > If the issue here is really with from/in, maybe we could make the concept of a function set a little more general?

My concern with this is that the property name `function_set` isn't obviously tied to indirect calls.  For example, people might assume the property guarantees that the functions reside in the same device image (which is not the case).

Reviewer 2: Yeah, I deliberately tried not to tie it to indirection.  It would just be defined as a declared set of functions that other properties could hook into to modify the behavior of all functions in the set.  Maybe that's too far.

Reviewer 3: Yeah, I'd rather not go that far.  We might add sets for some other purpose in the future, and I want to preserve our ability to have "indirect call sets" that are different from "some other sets".

Reviewer 2: Ok, I think we're at an impasse.  I'm still leaning towards to/in as the cleaner mental model.

Should we get a third opinion?

Reviewer 3: I only meant that we should not add the `function_set` property.  I'm OK with the to/in model where the set names are integrated into the "indirectly callable" property.  I think you previously proposed properties like this:

* `assume_indirect_calls`
* `assume_indirect_calls_to<set>`
* `indirectly_callable`
* `indirectly_callable_in<set>`

Refining this a bit, should we allow a pack of set names in `assume_indirect_calls_to`, making it:

* `assume_indirect_calls`
* `assume_indirect_calls_to<set1, set2, ...>`
* `indirectly_callable`
* `indirectly_callable_in<set>`

Reviewer 2: > Refining this a bit, should we allow a pack of set names in `assume_indirect_calls_to`, making it:
> 
> * `assume_indirect_calls`
> * `assume_indirect_calls_to<set1, set2, ...>`
> * `indirectly_callable`
> * `indirectly_callable_in<set>`

This looks good to me. 

Part of me wonders if `indirectly_callable_in` should also accept a pack of set names, so functions can be declared as members of multiple sets.  But I can't immediately think of a use-case, especially if sets are typically going to correspond to required aspects.

Author: Renamed the extension doc to reflect new names in b7cdc2d8e86b72368f332fc108faf92b751db205 and updated design doc with them in 46897b3337cadf7c7ba7ed7f66f81ee1e59e9f18

|The `calls_indirectly` property indicates that a SYCL kernel function is
performing calls through virtual member functions and declares use of one or
more sets of virtual member functions.

Optional parameter `SetIds` specifies which sets of virtual member functions are
declared to be used by a kernel, it must be zero or more C\++ typenames. If the
argument is omitted (zero C++ typenames specified), then a kernel is considered
to be using the default set of virtual member functions.

Calling a virtual member function, which does not belong to any of sets of
virtual member functions declared to be used is an undefined behavior.
|===

If a virtual member function is called from device code, both definition and
declaration of that function must be decorated with the `indirectly_callable`
property. `SetId` property parameter must match between definition and

> declaration, or otherwise behavior is undefined.

Author: Let me do another round of prototyping here. This wording is a result of my previous experiments, where both definition and declaration had to be annotated, or otherwise some functionality was missing (either proper LLVM IR codegen, or FE diagnostics). Perhaps I missed something an prototyped it wrong and I'd done before we have agreed that we can parse properties in FE. The latter changes the landscape a bit, so with it we may be in a bit better position overall.

> declaration, or otherwise behavior is undefined.

Author: Updated the wording to match #11861 in 4b622e169b8078305d130b0b02c4480b54f712b8

in another extensions.

Virtual member functions that are decorated with the `indirectly_callable`
property are considered to be device functions, i.e. they  must obey the
restrictions listed in section 5.4 of the core SYCL specification "Language
restrictions for device functions". Virtual member functions that are not
decorated with this attribute do not need to obey these restrictions, even if
other definitions of that virtual member function in other classes in the
inheritance hierarchy are decorated with the attribute.

[source,dpcpp]
----
using syclext = sycl::ext::oneapi::experimental;

struct set_A;
struct set_B;

class Foo {
public:
  // properties to functions should be applied using the macro:
  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_A) void foo() {}

  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_A) void bar() {};

  // both declaration and definition should be annotated:
  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_B) void baz();
};

// both declaration and definition should be annotated:
void SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_B) Foo::baz() {}

// kernel calling virtual function should also be annotated:

> /* */.single_task(syclext::properties{syclext::calls_indirectly<set_A>}, [=] {

Author: I agree, that looks better. Applied to all code snippets in the extension doc in 7db0a2aec63cdc0e51bd596ac6f539c2fbfb3a3b

  Foo *ptr = /* ... */;
  ptr->bar()

  // Note: this kernel can only call 'Foo::foo' and 'Foo::bar' but not
  // 'Foo::baz', because the latter is declared within a different set.
});
----

Reasons about why it may be necessary to put different virtual functions into

> different sets are explained in further sections, but for simplicity purposes

Reviewer 3: I noticed this too.  I think @AlexeySachkov either forgot to write that section, or he hasn't written it yet.

> different sets are explained in further sections, but for simplicity purposes

Author: Indeed, I forgot to add some info to explain it better. Tried to address this in 6507226724b4ae8293be656329f237567a8f865b

both properties can be used without explicitly specifying a set, thus using
the default set:

[source,dpcpp]
----
using syclext = sycl::ext::oneapi::experimental;

struct set_A;

class Foo {
public:
  // This virtual member function belongs to the default set of virtual
  // functions.
  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void foo() {}

  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_A) void bar() {}
};

/* */.single_task(syclext::properties{syclext::calls_indirectly<>}, [=] {
  Foo *ptr = /* ... */;
  ptr->bar()

  // Note: this kernel can only call 'Foo::foo' but not 'Foo::bar', because the
  // latter belongs to a different (non-default) set of virtual functions.
});
----

NOTE: `void` typename is explicitly reserved as identifier of the default
virtual functions set. It can still be used, but with the expectation that it

> is also implicitly used by the extension implementation.

Reviewer 3: This wording seems a little weird to me because "implicitly used by the extension implementation" makes me think that using the type `void` might somehow interfere with the implementation.  Can we say this instead?

> By convention, the type `void` is used to denote the default set of virtual functions.  When the typename is omitted from the `SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY` macro, the type `void` is used by default.  Applications may also explicitly use the type `void` to denote this default set of virtual functions.

Author: Thanks! I've applied the suggestion in a07048ffca79dec10df6257b2e7d396b4960c7b7

=== Optional kernel features handling

The core SYCL specification (5.8 Attributes for device code) says the following
in the description of `device_has` attribute for SYCL kernels and non-kernel
device functions.

When the attribute is applied to a kernel:

> \... it causes the compiler to issue a diagnostic if the kernel (or any of the
> functions it calls) uses an optional feature that is associated with an aspect
> that is not listed in the attribute.

When the attribute is applied to a function:

> \... it causes the compiler to issue a diagnostic if the device function (or
> any of the functions it calls) uses an optional feature that is associated
> with an aspect that is not listed in the attribute.

Due to dynamic nature of virtual member functions, compiler in general case is
not able to perform static analysis of a call graph in order to understand which
exact virtual functions are called from which kernels.

Instead, information from the new properties is used by an implementation to
issue such diagnostic. When determining a set of aspects which are used by a

> SYCL kernel function, implementation also takes into account all aspects which

Reviewer 2: ```suggestion
SYCL kernel function, an implementation must take into account all aspects which
```

Author: Applied in 683bc66e92e3acb4c652791cfd7a815e5afa5783

are used by all virtual member functions included into all sets of virtual
member functions declared to be used by a kernel.

[source,dpcpp]
----
using syclext = sycl::ext::oneapi::experimental;

struct set_fp64;
struct set_fp16;

struct Foo {
  // This function uses 'fp64' aspect
  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_fp64) void f64() {
    double d = 3.14;
  }

  // This function uses 'fp16' aspect
  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_fp16) void f16() {

>     sycl::half h = 2.71;

Reviewer 3: ```suggestion
    sycl::half h = 2.71f;
```

Otherwise, the kernel also implicitly uses `fp64`.

defined by the core SYCL specification.

[source,dpcpp]
----
using syclext = sycl::ext::oneapi::experimental;

struct set_A;
struct set_B;

struct Foo {
  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_A) void foo() {
    double d = 3.14;
  }

  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_B) void bar() {}
};

sycl::queue q(/* device selector returns a device *without* fp64 support */);
assert(!q.get_device().has(sycl::aspect::fp64));

q.single_task(syclext::properties{syclext::calls_indirectly<set_A>}, [=] {
  // Exception is expected to be thrown, because target device doesn't support
  // fp64 aspect and it is required by 'Foo::foo' which is included into 'set_A'
});

q.single_task(syclext::properties{syclext::calls_indirectly<set_B>}, [=] {
  // No exceptions are expected, because 'set_B' doesn't bring any requirements
  // for optional kernel features.
});
----

An implementation may not raise a compile time diagnostic or a run time
exception merely due to speculative compilation of a virtual member function for
a device when the application does not specify a use of virtual member functions

> through the correponding properties.

Author: Applied in 683bc66e92e3acb4c652791cfd7a815e5afa5783

[source,dpcpp]
----
using syclext = sycl::ext::oneapi::experimental;

struct Foo {
virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void foo() {
  double d = 3.14;
}

virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void bar() {}
};

sycl::queue q(/* device selector choosing a device *without* fp64 support */);
assert(!q.get_device().has(sycl::aspect::fp64));

auto *Storage = sycl::malloc_device<Foo>(1, q);

q.single_task([=] {
  // The kernel is not submitted with 'calls_indirectly' property and therefore
  // it is not considered to be using any of virtual member functions of 'Foo'.
  // This means that the object of 'Foo' can be successfully created by this
  // kernel, regardless of whether a target device supports 'fp64' aspect which
  // is used by 'Foo::foo'.
  // No exceptions are expected to be thrown.
  new (Storage) Foo;
});
----

==== Interaction with `reqd_sub_group_size` attribute

The `reqd_sub_group_size` attribute is a bit of a special case comparing to
other optional kernel features, because it requires to compile a kernel in a
certain way, which may require special handling for all functions which are
called from it.

> When the same helper function is called from two or more kernels

Reviewer 2: I think this is the first occurrence of "helper function", and I'm not sure what it means.

Author: It was intended as a reference to the function from the previous paragraph, which is called from two kernels with different `reqd_sub_group_size` attributes. I've rewrote the paragraph to avoid using that undefined reference in 11f3a7d520fc29174fd21055704ecc97bd051c12

with different `reqd_sub_group_size` attribute, it may be required for the
implementation to duplicate that helper function to create different versions
of it tailored to different sub-group sizes. It can be done in a straightforward
manner when operating on a static call graph.

Virtual member functions are essentially called indirectly and pointers to them
are initialized just once when an object of a polymorhpic class is being

> created. Therefore, to support calling such virtual member function from two or

Reviewer 2: I think this explanation should appear earlier in the specification.  It explains why virtual member functions are described as being "indirectly callable", rather than having a dedicated property for virtual member functions.

Author: I have attempted to improve "intro" in the `New properties` section to address this, see 65016fc0248ee61b753d7aa479e0a67774b8ca5f

more kernels with different `reqd_sub_group_size`, each kernel may need to

> receive a different pointer to a different version of a virtual member function.

Reviewer: I'm not sure that I'm reading this correctly. A polymorphic object contains a single pointer to the virtual function table associated with the dynamic type of the class (in the case of multiple inheritance, there may be more pointers, but that doesn't matter here). The virtual function table is static and built at compile time. When a polymorphic object is constructed, its (single) virtual function table pointer is updated as construction begins for each base class progressing through the most derived class. So it isn't correct to say that "pointers to them
are initialized just once when an object of a polymorphic class is being created." I don't see what `reqd_sub_group_size` has to do with this.

Author: `reqd_sub_group_size` attribute is effectively lowered into vectorization width for a kernel. The main question here is what happens when we have several kernels with different `reqd_sub_group_size` attribute calling the same virtual function? There has to be some sort of contract within the compiler so that all kernels (which are vectorized differently) can perform call to the same function (which may be vectorized in some other way distinct from all kernels).

Moreover, when a function is called from a kernel with certain `reqd_sub_group_size`, then it is expected that all sub-group built-ins will properly work within it, like querying sub-group size should return the right value regardless of which kernel the function was called from.

To avoid possibly posing such multi-versioning requirements on implementations,
virtual member functions can only be called from kernels with _primary_
sub-group-size as defined by
link:../proposed/sycl_ext_oneapi_named_sub_group_sizes.asciidoc[
sycl_ext_oneapi_named_sub_group_sizes] extension, or otherwise behavior is

> undefined.

Author: Ok, my concern still stands: because we don't have a way to set reqd-sub-group size for a non-kernel function at SPIR-V level, we can't really annotate them at SYCL level. This is more of an implementation issue than a language issue, but I've recorded it in 95fd59fa592e9c66a7c3352d9771252130f68d04 to improve the situation later.

NOTE: for implementations that don't support
`sycl_ext_oneapi_named_sub_group_sizes` extension, virtual member functions can
only be called from kernels which *don't* have `reqd_sub_group_size` attribute
set on them explicitly, or otherwise behavior is undefined.

=== Kernel bundles and device images

When an object of a polymorphic class is constructed, it stores a pointer to
virtual table, which points to its virtual member functions. Addresses of those
functions are accessible and valid only within a kernel bundle containing a
kernel which used to construct an object.

Performing calls to virtual member functions of an object constructed in a

> kernel from a different kernel bundle is an undefined behavior.

Reviewer 3: This limitation seems surprising to me.  If this is really a limitation, what happens if the user does *not* use the kernel bundle APIs but compiles the program with `-fsycl-device-code-split=per_kernel`?  This will cause every kernel to be in its own bundle.  If the application creates the object in one kernel and then uses the object in another kernel, won't we have this same situation where an object constructed in one bundle is used in another?

Reviewer 3: > My expectation is that it would just work, because implementation will merge all dependent kernels and virtual functions together based on the new properties. But I must admit, that it is not spelled out anywhere in the spec.

I think you are suggesting that `sycl-post-link` will always bundle together into the same device image a kernel that constructs a polymorphic object along with all the kernels that could call virtual methods on that polymorphic object.  I think this could work, but I think it will get tricky.  Consider this example:

```
using syclext = sycl::ext::oneapi::experimental;

struct set1;
struct set2;

struct Foo {
  virtual
  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::indirectly_callable<set1>)
  void foo() {}

  virtual
  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::indirectly_callable<set2>)
  void bar() {}
};

int main() {
  sycl::queue q;

  auto *Storage = sycl::malloc_device<Foo>(1, q);

  q.single_task([=] {  // Kernel A
    new (Storage) Foo;
  });

  if (q.get_device().has(sycl::aspect::fp64)) {
    auto props = syclext::properties{syclext::calls_indirectly<set1>};
    q.single_task(props, [=] {  // Kernel B
      double d = 3.14;
      Storage->foo();
    });
  }

  auto props = syclext::properties{syclext::calls_indirectly<set2>};
  q.single_task(props, [=] {  // Kernel C
    Storage->bar();
  });
}
```

In this example kernels A, B and C may all need to be bundled together in order to ensure that the allocation of object `Foo` is in the same device image as the calls on its virtual member functions.  Will you be able to determine this because you will see the construction of `Foo` in kernel A and calls to virtual member functions in kernels B and C?

You will also have to *conditionally* bundle kernel B into that device image only if the device supports the `fp64` feature.  This can be done via dynamic linking, but it's a different use of dynamic linking beyond what you have described in the design document.

If this is too complicated, we could shift some of the responsibility to the user, and require the user to decorate kernel A with a property indicating the "set" that it belongs to.  It would then be the user's responsibility to ensure that the same "set" is used for the kernel that allocates the object and also for the kernels that calls virtual functions on the object.

BTW, a lot of the complexity comes from the fact that we allow one kernel to allocate a polymorphic object (via placement new) and then make virtual function calls on that object from a different kernel.  Do we know that this is actually an important requirement?  If we restricted virtual functions to the case when the object is allocated in the same kernel as the calls, the design would be simpler.

Author: > I think you are suggesting that `sycl-post-link` will always bundle together into the same device image a kernel that constructs a polymorphic object along with all the kernels that could call virtual methods on that polymorphic object.

Not necessarliy. My plan was, in fact, that runtime will be responsible for bundling all necessary kernels and virtual functions together - unless it is already explicitly done by user through explicit use of `kernel_bundle` API.

> BTW, a lot of the complexity comes from the fact that we allow one kernel to allocate a polymorphic object (via placement new) and then make virtual function calls on that object from a different kernel. Do we know that this is actually an important requirement?

We have been provided with the following application as basic acceptance example: [Exawind/exw-virtuals-app](https://github.com/Exawind/exw-virtuals-app) and it constructs a polymorphic object and uses its virtual member functions in separate kernels.

Reviewer 3: > Not necessarliy. My plan was, in fact, that runtime will be responsible for bundling all necessary kernels and virtual functions together - unless it is already explicitly done by user through explicit use of `kernel_bundle` API.

When you say "runtime will be responsible for bundling all necessary kernels and virtual functions together", I assume you mean that the runtime will construct a single Level Zero or OpenCL module that contains all these kernels.  Right?  How can the runtime do this, though, in the AOT case?  In the AOT case, the kernels are already bundled into device images that have been AOT-compiled to native code.  We do not have any way at runtime to combine these native code modules together into a single module.

> We have been provided with the following application as basic acceptance example: [Exawind/exw-virtuals-app](https://github.com/Exawind/exw-virtuals-app) and it constructs a polymorphic object and uses its virtual member functions in separate kernels.

I was thinking about this more, and I think there is a problem even in a simple case.  Consider this simple case:

```
using syclext = sycl::ext::oneapi::experimental;

struct Foo {
  virtual
  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::indirectly_callable<>)
  void bar() {}
};

int main() {
  sycl::queue q;

  auto *Storage = sycl::malloc_device<Foo>(1, q);
  q.single_task([=] {  // Kernel A
    new (Storage) Foo;
  });

  // Many things happen ...

  auto props = syclext::properties{syclext::calls_indirectly<>};
  q.single_task(props, [=] {  // Kernel B
    Storage->bar();
  });
}
```

Let's assume the user is JIT-compiling and that there is a single SPIR-V module that contains kernels A and B.  When kernel A is JIT compiled, we cache a copy of the native code module.  The `Storage` object ends up embedding the address of "bar" as part of the v-table.

In my example, kernel B runs a long time after kernel A.  By the time this happens, the native code module may have been evicted from the cache.  Therefore, when kernel B runs we may JIT compile the SPIR-V module again.  However, I assume that the address of "bar" might be different in this new module.  Thus, it will be a problem when kernel B tries to use the v-table of the object in `Storage`.

I can think of other cases that will cause kernel B to be re-JIT-compiled too.  For example, if kernel B contains a specialization constant.

Author: I'm less concerned about in-memory cache, because we literally had only one case when we had to evict it and that is SYCL CTS for specialization constants, which is quite heavy. In any case, I've added some notes about it and specialization constants handling in eb5ea4d3dfbaae21c8df1629a163673bd7e423b1. I'm not sure if the suggested way of dealing with specialization constants is good enough, so I've also recorded it as an issue in 3728caeadfb218e8dc4370ae56d1e909a18ce9e4

Back to overall design:

> When you say "runtime will be responsible for bundling all necessary kernels and virtual functions together", I assume you mean that the runtime will construct a single Level Zero or OpenCL module that contains all these kernels. Right?

Right

> How can the runtime do this, though, in the AOT case?

In AOT case we should know target capabilities and therefore all the bundling will happen at compile-time. I.e. every device image that uses virtual functions is expected to be self-contained already as it is now.

With JIT flow, we don't know target capabilities in advance and all the bundling will happen at runtime as described in the design doc.

[source,dpcpp]
----
using syclext = sycl::ext::oneapi::experimental;

struct Base {
  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void foo() {}
};

class Constructor;
class Use;

int main() {
  sycl::queue Q;

  Base *Obj = sycl::malloc_device<Base>(1, Q);
  int *Result = sycl::malloc_shared<int>(2, Q);

  auto bundleA
    = sycl::get_kernel_bundle<sycl::bundle_state::executable>(Q.get_context(),
        {sycl::get_kernel_id<Constructor>()});
  auto bundleB
    = sycl::get_kernel_bundle<sycl::bundle_state::executable>(Q.get_context(),
        {sycl::get_kernel_id<Use>()});

  Q.submit([&](sycl::handler &CGH) {
    CGH.use_kernel_bundle(bundleA);
    CGH.single_task<Constructor>([=] {
      // Only placement new can be used within device functions.

>       new (Obj) Derived;

Reviewer 4: Should the definition of `Derived` class be added to this code snippet?

Author: Right, it should be `Base`, fixed in 683bc66e92e3acb4c652791cfd7a815e5afa5783

      // constructed within kernel bundle `bundleA`
      Obj->foo();
    });
  });

  return 0;
}
----

== Example usage

[source,dpcpp]
----
#include <sycl/sycl.hpp>

using syclext = sycl::ext::oneapi::experimental;

class Base {
public:
  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY()
  int get_random_number() {
    return 4; // Chosen by fair dice roll. Guaranteed to be random
  }

  // Not considered to be a device function, can use full set of C++ features
  virtual int get_host_random_number() {
    throw std::runtime_error("Not Implemented");
  }
};

class Derived : public Base {
public:
  SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY()
  int get_random_number() override {
    return 221;
  }
};

int main() {
  sycl::queue Q;

  Base *Obj = sycl::malloc_device<Derived>(1, Q);
  int *Result = sycl::malloc_shared<int>(1, Q);

  Q.single_task([=] {
    // Only placement new can be used within device functions.
    new (Obj) Derived;
  });

  auto props = syclext::properties{syclext::calls_indirectly<>};
  Q.single_task(props, [=] {
    Base B;
    Result[0] = B.get_random_number();
  }).wait();

>   assert(A[0] == 4);

Reviewer 3: ```suggestion
  assert(Result[0] == 4);
```

There is no variable `A`, so I think this is what you intended?  Same for the other occurrences of `A` below.
